src/opm/parser/eclipse/EclipseState/Schedule/WellProductionProperties.cpp

```diff
@@ -43,50 +44,64 @@ namespace Opm {
+        if ( record.getItem( "THP" ).hasValue(0) )
+            p.THPH = record.getItem("THP").getSIDouble(0);
 
-            // when there is an effective control mode specified by WHISTCL, we always use this control mode
-            if (effectiveHistoryProductionControl(controlModeWHISTCL) ) {
-                cmode = controlModeWHISTCL;
-            }
+        const auto& cmodeItem = record.getItem("CMODE");
+        if ( cmodeItem.defaultApplied(0) ) {
+            const std::string msg = "control mode can not be defaulted for keyword WCONHIST";
+            throw std::invalid_argument(msg);
+        }
 
-            if (effectiveHistoryProductionControl(cmode) ) {
-                p.addProductionControl( cmode );
-                p.controlMode = cmode;
-            } else {
-                const std::string cmode_string = cmodeItem.getTrimmedString( 0 );
-                const std::string msg = "unsupported control mode " + cmode_string + " for WCONHIST";
-                throw std::invalid_argument(msg);
-            }
+        namespace wp = WellProducer;
+        auto cmode = wp::ControlModeFromString( cmodeItem.getTrimmedString( 0 ) );
 
-            // always have a BHP control/limit, while the limit value needs to be determined
-            // the control mode added above can be a BHP control or a type of RATE control
-            if ( !p.hasProductionControl( wp::BHP ) )
-                p.addProductionControl( wp::BHP );
+        // when there is an effective control mode specified by WHISTCL, we always use this control mode
+        if (effectiveHistoryProductionControl(controlModeWHISTCL) ) {
+            cmode = controlModeWHISTCL;
+        }
 
+        if (effectiveHistoryProductionControl(cmode) ) {
+            p.addProductionControl( cmode );
+            p.controlMode = cmode;
+        } else {
+            const std::string cmode_string = cmodeItem.getTrimmedString( 0 );
+            const std::string msg = "unsupported control mode " + cmode_string + " for WCONHIST";
+            throw std::invalid_argument(msg);
+        }
 
-            if (cmode == wp::BHP)
-                p.BHPLimit = record.getItem( "BHP" ).getSIDouble( 0 );
-            else
+        // always have a BHP control/limit, while the limit value needs to be determined
+        // the control mode added above can be a BHP control or a type of RATE control
+        if ( !p.hasProductionControl( wp::BHP ) )
+            p.addProductionControl( wp::BHP );
+
+        if (cmode == wp::BHP) {
+            p.BHPLimit = p.BHPH;
+        } else {
+            // when first time the well is claimed to be history matching producer
+            // or switching from injector to producer
+            // or switching from BHP control to RATE control
+            // we use the defaulted BHP limit, otherwise, we use the previous BHP limit
+            if ( prev_properties.predictionMode || switching_from_injector
```

Comment:
What if we switch back and forth between injector and producer several times - so this is not the first time we execute code with `switching_from_injector == true`? Maybe it would be a better solution to just have the `BHPLimit` default initialized to 1 `atm` in the header file?

Reply:
It does not care how many times you switch back and forth. As long as the switching happen, it is the same. The BHP limit from the injector should not be used for producer at all. 

> Maybe it would be a better solution to just have the BHPLimit default initialized to 1 atm in the header file?

possibly. let me check.

Reply:
> Maybe it would be a better solution to just have the BHPLimit default initialized to 1 atm in the header file?

It is possible, then we need to figure out when we keep it or not, then there will be some `NOT` logic, which is not necessarily more clear.  If it is okay, I will prefer not to change it.